core/server/master/src/test/java/alluxio/master/file/meta/UfsSyncCachePathTest.java

```diff
@@ -0,0 +1,247 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.meta;
+
+import alluxio.file.options.DescendantType;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+public class UfsSyncCachePathTest {
+
+  private String mGrandParentDir;
+  private String mParentPath;
+  private String mChildPath;
+  private String mChildFile;
+  private UfsSyncPathCache mUspCache;
+
+  @Before
+  public void before() throws Exception {
+    mGrandParentDir = "/dir1";
+    mParentPath = "/dir1/dir2";
+    mChildPath = "/dir1/dir2/dir3";
+    mChildFile = "/dir1/dir2/file";
+    mUspCache = new UfsSyncPathCache();
+  }
+
+  @Test
+  public void ignoreIntervalTime() {
+    // request from getFileInfo
+    boolean shouldSync = mUspCache.shouldSyncPath(mParentPath, -1, true);
+    Assert.assertFalse(shouldSync);
+    shouldSync = mUspCache.shouldSyncPath(mParentPath, 0, true);
+    Assert.assertTrue(shouldSync);
+    // request from listStatus
+    shouldSync = mUspCache.shouldSyncPath(mParentPath, -1, false);
+    Assert.assertFalse(shouldSync);
+    shouldSync = mUspCache.shouldSyncPath(mParentPath, 0, false);
+    Assert.assertTrue(shouldSync);
+  }
+
+  /**
+   * The path itself is in UfsSyncCachePath.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void getFileInfoInCache() throws Exception {
+    getFileInfoInCache(DescendantType.NONE);
+    getFileInfoInCache(DescendantType.ONE);
+    getFileInfoInCache(DescendantType.ALL);
+  }
+
+  private void getFileInfoInCache(DescendantType descendantType) throws Exception {
+    mUspCache.notifySyncedPath(mParentPath, descendantType);
+    Thread.sleep(100);
+    // request from getFileInfo
+    boolean shouldSync = mUspCache.shouldSyncPath(mParentPath, 60, true);
+    Assert.assertTrue(shouldSync);
+    shouldSync = mUspCache.shouldSyncPath(mParentPath, 200, true);
+    Assert.assertFalse(shouldSync);
+  }
+
+  /**
+   * The direct parent dir of path is in UfsSyncCachePath.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void getFileInfoFromDirectParent() throws Exception {
+    getFileInfoFromDirectParent(DescendantType.NONE);
+    getFileInfoFromDirectParent(DescendantType.ONE);
+    getFileInfoFromDirectParent(DescendantType.ALL);
+  }
+
+  private void getFileInfoFromDirectParent(DescendantType descendantType) throws Exception {
+    mUspCache.notifySyncedPath(mParentPath, descendantType);
+    Thread.sleep(100);
+    // test child directory
+    boolean shouldSync = mUspCache.shouldSyncPath(mChildPath, 60, true);
+    Assert.assertTrue(shouldSync);
+    shouldSync = mUspCache.shouldSyncPath(mChildPath, 200, true);
+    Assert.assertFalse(shouldSync);
```

Comment:
Before this PR, this would have returned false for timeout and true in time, because getFileInfo is the same for file and directory.  The listStatus and other methods should consider the direct parent directory specially.